sdk/go/pulumi/context.go

```diff
@@ -531,6 +617,13 @@ func (ctx *Context) getOpts(opts ...ResourceOpt) (URN, []URN, bool, string, bool
 		}
 	}
 
+	providers := mergeProviders(t, opts...)
```

Comment:
Feels like we should just pull this off of `*ResourceState` rather than recomputing it.

Reply:
Done. Thanks for pointing this out. I imagine the concern here is perf related, as piping through the additional parameter through two functions is a little ugly? Looking more closely, we're computing a set of sync and async properties, and we should probably have `prepareOpts` functions for each to make this a little cleaner.

Reply:
(It's been a minute since I had to worry at object allocation level perf)